internal/db/db.go

```diff
@@ -51,10 +52,108 @@ func createOwnerInfoTable() error {
 	return nil
 }
 
-func FetchVoucher(guid []byte) (Voucher, error) {
+// FetchVoucher doesn't take into account mfg_voucher where go-fdo stores vouchers that have not been extended yet
+// we don't need it right now, but for use cases where manufacturers just initializes empty devices (tpm) this is going to be needed.
+//
+// FetchVoucher returns a single voucher filtered by provided fields.
+// Supported filters (keys):
+// - "guid" (expects []byte)
+// - "device_info" (expects string)
+// If more than one voucher matches, an error is returned.
+// Note: This does not query mfg_voucher (unextended vouchers).
+func FetchVoucher(filters map[string]interface{}) (Voucher, error) {
 	var voucher Voucher
-	err := db.QueryRow("SELECT guid, cbor FROM owner_vouchers WHERE guid = ?", guid).Scan(&voucher.GUID, &voucher.CBOR)
-	return voucher, err
+	if len(filters) == 0 {
+		return voucher, fmt.Errorf("no filters provided")
+	}
+
+	query := "SELECT guid, device_info, cbor, created_at, updated_at FROM owner_vouchers WHERE 1=1"
+	args := make([]interface{}, 0, 2)
+
+	if v, ok := filters["guid"]; ok {
+		b, ok := v.([]byte)
+		if !ok {
+			return voucher, fmt.Errorf("invalid type for guid filter; want []byte")
+		}
+		query += " AND guid = ?"
+		args = append(args, b)
+	}
+	if v, ok := filters["device_info"]; ok {
+		s, ok := v.(string)
+		if !ok {
+			return voucher, fmt.Errorf("invalid type for device_info filter; want string")
+		}
+		query += " AND device_info = ?"
+		args = append(args, s)
+	}
+
+	// Fetch up to two rows to detect duplicates
+	query += " ORDER BY updated_at DESC LIMIT 2"
+
+	rows, err := db.Query(query, args...)
+	if err != nil {
+		return voucher, err
+	}
+	defer func() { _ = rows.Close() }()
+
+	if !rows.Next() {
+		return voucher, sql.ErrNoRows
+	}
+	var createdInt, updatedInt int64
+	if err := rows.Scan(&voucher.GUID, &voucher.DeviceInfo, &voucher.CBOR, &createdInt, &updatedInt); err != nil {
+		return voucher, err
+	}
+	voucher.CreatedAt = time.UnixMicro(createdInt)
+	voucher.UpdatedAt = time.UnixMicro(updatedInt)
+
+	// If there is a second row, error out due to ambiguity
+	if rows.Next() {
+		return Voucher{}, fmt.Errorf("multiple vouchers matched filters")
+	}
+
+	return voucher, nil
+}
+
+// QueryVouchers returns all owner vouchers matching optional filters (without CBOR), ordered by updated_at DESC.
+func QueryVouchers(filters map[string]interface{}) ([]Voucher, error) {
```

Comment:
Same here, lot of duplicated code: `FetchVoucher` is pretty much the same function as `QueryVouchers`. I suggest to redefine the `QueryVouchers` function so it takes an additional boolean argument to determine if the CBOR field must be included in the resulting vouchers or not and call it from `FetchVoucher` and there check if the result's size is greater than 1 to detect duplicates.

Reply:
right, I think we don't even need the cbor boolean as omitempty seems enough (and we don't populate that in QueryVouchers)